source/CustomSale.cpp

```diff
@@ -0,0 +1,368 @@
+			relativeOffsets.emplace(it.first, it.second);
+		else
+			ours->second += it.second;
+	}
+	// Same thing for the outfits.
+	for(const auto &it : other.relativeOutfitPrices)
+	{
+		auto ours = relativeOutfitPrices.find(it.first);
+		if(ours == relativeOutfitPrices.end())
+			relativeOutfitPrices.emplace(it.first, it.second);
+		else if(ours->second < it.second)
+			ours->second = it.second;
+	}
+	for(const auto &it : other.relativeOutfitOffsets)
+	{
+		auto ours = relativeOutfitOffsets.find(it.first);
+		if(ours == relativeOutfitOffsets.end())
+			relativeOutfitOffsets.emplace(it.first, it.second);
+		else
+			ours->second += it.second;
+	}
+	return true;
+}
+
+
+
+double CustomSale::GetRelativeCost(const Outfit &item) const
+{
+	// Outfit prices have priority over outfitter prices, so consider them first,
+	// and only consider the outfitter prices if the outfits have no set price.
+	const auto &baseRelative = relativeOutfitPrices.find(&item);
+	double baseRelativePrice = (baseRelative != relativeOutfitPrices.cend() ? baseRelative->second : DEFAULT);
+	if(baseRelativePrice == DEFAULT)
+		for(const auto &it : relativePrices)
+			if(it.first->Has(&item))
+			{
+				baseRelativePrice = it.second;
+				break;
+			}
+	const auto &baseOffset = relativeOutfitOffsets.find(&item);
+	double baseOffsetPrice = (baseOffset != relativeOutfitOffsets.cend() ? baseOffset->second : DEFAULT);
+	for(const auto &it : relativeOffsets)
+		if(it.first->Has(&item))
+		{
+			if(baseOffsetPrice == DEFAULT)
+				baseOffsetPrice = 0.;
+			baseOffsetPrice += it.second;
+		}
+	// Apply the relative offset on top of each others, the result being applied on top of the relative price.
+	// This means that an outfit can be affected by an outfitter offset, a custom outfit price, and outfit prices.
+	if(baseRelativePrice != DEFAULT)
+		return baseRelativePrice + (baseOffsetPrice != DEFAULT ? baseRelativePrice * baseOffsetPrice : 0.);
+	else if(baseOffsetPrice != DEFAULT)
+		return 1. + baseOffsetPrice;
+	else
+		return 1.;
+}
+
+
+
+CustomSale::SellType CustomSale::GetSellType() const
+{
+	return sellType;
+}
+
+
+
+const string &CustomSale::GetShown(CustomSale::SellType sellType)
+{
+	return show.find(sellType)->second;
+}
+
+
+
+const Sale<Outfit> CustomSale::GetOutfits() const
+{
+	Sale<Outfit> seen;
+	for(auto it : relativeOutfitPrices)
+		seen.insert(it.first);
+	for(auto it : relativeOutfitOffsets)
+		seen.insert(it.first);
+	for(auto &&sale : relativePrices)
+		seen.Add(*sale.first);
+	for(auto &&sale : relativeOffsets)
+		seen.Add(*sale.first);
+	return seen;
+}
+
+
+
+bool CustomSale::Has(const Outfit &item) const
+{
+	for(auto it : relativeOutfitPrices)
+		if(it.first == &item)
+			return true;
+	for(auto it : relativeOutfitOffsets)
+		if(it.first == &item)
+			return true;
+	for(auto &&sale : relativePrices)
+		if(sale.first->Has(&item))
+			return true;
+	for(auto &&sale : relativeOffsets)
+		if(sale.first->Has(&item))
+			return true;
+	return false;
+}
```

Comment:
I dont think we ever do sorting and the like? Idk

Reply:
Aren't these all `std::map<const Outfit *, [something]>`?
That means they're sorted by the value of the Outfit pointers, so doing `find(&item)` can perform a binary search for you, instead of the probably slower linear search you've got going here.

Reply:
I can't do it for the Sale<outfit> but I'll do it for the const outfit*